airbyte-cdk/bulk/toolkits/load-object-storage/src/main/kotlin/io/airbyte/cdk/load/file/object_storage/ProtoToCsvFormatter.kt

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (c) 2025 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.cdk.load.file.object_storage
+
+import io.airbyte.cdk.load.command.DestinationStream
+import io.airbyte.cdk.load.data.ObjectType
+import io.airbyte.cdk.load.data.collectUnknownPaths
+import io.airbyte.cdk.load.data.csv.toCsvHeader
+import io.airbyte.cdk.load.data.withAirbyteMeta
+import io.airbyte.cdk.load.file.csv.toCsvPrinterWithHeader
+import io.airbyte.cdk.load.message.DestinationRecordProtobufSource
+import io.airbyte.cdk.load.message.DestinationRecordRaw
+import io.airbyte.cdk.load.message.Meta
+import io.airbyte.protocol.models.v0.AirbyteRecordMessageMetaChange
+import java.io.OutputStream
+
+class ProtoToCsvFormatter(
+    stream: DestinationStream,
+    private val outputStream: OutputStream,
+    rootLevelFlattening: Boolean,
+    extractedAtAsTimestampWithTimezone: Boolean,
+) : ObjectStorageFormattingWriter {
+
+    private val finalSchema: ObjectType = stream.schema.withAirbyteMeta(rootLevelFlattening)
+
+    private val header: Array<String> = finalSchema.toCsvHeader()
+    private val printer = finalSchema.toCsvPrinterWithHeader(outputStream)
+    private val unknownColumns = stream.schema.collectUnknownPaths()
+    private val unknownColumnsPresent = unknownColumns.isNotEmpty()
+
+    private val writer =
+        ProtoToCsvWriter(
+            header = header,
+            columns = stream.airbyteValueProxyFieldAccessors,
+            rootLevelFlattening = rootLevelFlattening,
+            extractedAtTsTz = extractedAtAsTimestampWithTimezone,
+        )
+
+    override fun accept(record: DestinationRecordRaw) {
+        val src = record.rawData
+        require(src is DestinationRecordProtobufSource) {
+            "ProtoToCsvFormatter only supports DestinationRecordProtobufSource"
+        }
+        val changes = record.rawData.sourceMeta.changes.toMutableList()
+        if (unknownColumnsPresent) {
+            unknownColumns.forEach {
```

Comment:
nit: what if we precomputed this changes list in the constructor? something like
```kotlin
private val unknownColumnChanges = unknownColumns.map { Meta.Change(...) }

override fun accept() {
  // ...
  printer.printRecord(*writer.toCsvRow(record, src, record.rawData.sourceMeta.changes + unknownColumnChanges))
}
```